ext/drx/drx.c

```diff
@@ -2429,3 +2441,363 @@ drx_expand_scatter_gather(void *drcontext, instrlist_t *bb, OUT bool *expanded)
+                                          &allow_for_unknown_instr_count);
+                            break;
+                        }
+                    }
+                }
+            }
+            if (allow_for_unknown_instr_inc(&allow_for_unknown_instr_count))
+                break;
+            advance_state(&detect_state, DRX_DETECT_RESTORE_AVX512_GATHER_EVENT_STATE_0,
+                          &allow_for_unknown_instr_count);
+            break;
+        case DRX_DETECT_RESTORE_AVX512_GATHER_EVENT_STATE_8:
+            if (instr_get_opcode(inst) == OP_kmovw) {
+                opnd_t dst0 = instr_get_dst(inst, 0);
+                if (opnd_is_reg(dst0) && opnd_get_reg(dst0) == DR_REG_K0) {
+                    opnd_t src0 = instr_get_src(inst, 0);
+                    if (opnd_is_reg(src0)) {
+                        reg_id_t tmp_gpr = opnd_get_reg(src0);
+                        if (reg_is_gpr(tmp_gpr)) {
+                            if (restore_scratch_mask_start <= info->raw_mcontext->pc &&
+                                info->raw_mcontext->pc <= prev_pc) {
+                                /* The scratch mask is always k0. this is hard-coded
+                                 * in drx. We carefully only update the lowest 16 bits
+                                 * because the mask was saved with kmovw.
+                                 */
+                                ASSERT(sizeof(info->mcontext->opmask[0]) ==
+                                           sizeof(long long),
+                                       "internal error: unexpected opmask slot size");
+                                info->mcontext->opmask[0] &= ~0xffffLL;
+                                info->mcontext->opmask[0] |=
+                                    reg_get_value(gpr_save_scratch_mask, info->mcontext) &
+                                    0xffff;
+                                advance_state(
+                                    &detect_state,
+                                    DRX_DETECT_RESTORE_AVX512_GATHER_EVENT_STATE_0,
+                                    &allow_for_unknown_instr_count);
+                            }
+                        }
+                    }
+                }
+            }
+            if (allow_for_unknown_instr_inc(&allow_for_unknown_instr_count))
+                break;
+            advance_state(&detect_state, DRX_DETECT_RESTORE_AVX512_GATHER_EVENT_STATE_0,
+                          &allow_for_unknown_instr_count);
+            break;
+        default: ASSERT(false, "internal error: invalid state.");
+        }
+    }
+}
+
+static bool
+drx_event_restore_state(void *drcontext, bool restore_memory,
+                        dr_restore_state_info_t *info)
+{
+    instr_t inst;
+    if (info->fragment_info.cache_start_pc == NULL)
+        return true; /* fault not in cache */
+    instr_init(drcontext, &inst);
+    byte *pc = decode(drcontext, dr_fragment_app_pc(info->fragment_info.tag), &inst);
```

Comment:
Shouldn't this *not* be done for !app_code_consistent?

Reply:
I've already fixed that, I am returning false now to fail the translation, since we can't determine safely whether this is an emulation sequence, so it may or may not be one. Resolving, re-open if necessary.

Reply:
> I've already fixed that

You mean I was looking at an old diff?

Reply:
No, I just meant I had changed that already before I read the comment. Ignore my comment, I will update.